File: examples/elect.c
/* Live Elect through the C FFI.
 *
 * Keep-operating refuses. Commit waits. A live canceler can stop the
 * ceremony. After the wait, commit installs the named heir.
 *
 * cc -I crates/sociacl-c/include examples/elect.c -L target/debug -lsociacl -o target/sociacl-elect-c
 * LD_LIBRARY_PATH=target/debug target/sociacl-elect-c
 */

#include "sociacl.h"

#include <stdio.h>
#include <stdlib.h>

static void die(const char *msg) {
    fprintf(stderr, "%s\n", msg);
    exit(1);
}

static const char *HEIR =
    "will heir-doc for object doc\n"
    "written-by alice\n"
    "cancelable-by executor\n"
    "discover heir bob\n";

int main(void) {
    sociacl_plane *plane = sociacl_plane_new();
    if (!plane) {
        die("plane");
    }
    sociacl_add_person(plane, "alice");
    sociacl_add_person(plane, "bob");
    sociacl_add_person(plane, "executor");
    sociacl_add_object(plane, "doc", "alice");

    char reason[256];
    if (sociacl_write_will(plane, HEIR, reason, sizeof reason) != 0) {
        die(reason);
    }
    if (sociacl_elect(plane, "doc", reason, sizeof reason) == 0) {
        die("live owner must refuse elect");
    }
    printf("elect while alice is live: %s\n", reason);

    sociacl_set_authn(plane, "alice", "gone");
    if (sociacl_elect(plane, "doc", reason, sizeof reason) != 0) {
        die(reason);
    }
    printf("elect started: %s\n", reason);

    if (sociacl_commit_elect(plane, "doc", reason, sizeof reason) == 0) {
        die("commit before wait must fail");
    }
    printf("commit before wait: %s\n", reason);

    uint64_t now = 0;
    sociacl_now(plane, &now);
    sociacl_set_now(plane, now + 10);
    if (sociacl_commit_elect(plane, "doc", reason, sizeof reason) != 0) {
        die(reason);
    }
    printf("commit after wait: %s\n", reason);

    int allowed = sociacl_check(
        plane, "read", "doc", "bob", "owner", reason, sizeof reason
    );
    printf("bob owns doc: allowed=%d reason=%s\n", allowed, reason);

    sociacl_plane_free(plane);
    return allowed == 1 ? 0 : 1;
}

File: examples/elect.rs
//! Live Elect ceremony. Keep-operating refuses. A live canceler can stop
//! a pending Elect. Commit installs only after the wait.
//!
//! Run: `cargo run -p sociacl-core --example elect`

use sociacl_core::{AuthnState, Plane, PredicateId, VerbError};

fn main() {
    let mut plane = Plane::new();
    let alice = plane.add_person("alice").id;
    let bob = plane.add_person("bob").id;
    let executor = plane.add_person("executor").id;
    let doc = plane.add_object("doc", &alice).id;
    let desk = plane.add_object("desk", &alice).id;
    plane
        .write_will_src(
            "will heir-doc for object doc\nwritten-by alice\ncancelable-by executor\ndiscover heir bob\n",
        )
        .unwrap();
    plane
        .write_will_src(
            "will heir-desk for object desk\nwritten-by alice\ncancelable-by executor\ndiscover heir bob\n",
        )
        .unwrap();

    match plane.elect(&doc) {
        Err(VerbError::KeepOperatingSuffices(_)) => {
            println!("elect while alice is live: keep-operating suffices")
        }
        other => panic!("expected keep-operating refuse, got {other:?}"),
    }

    plane.set_authn(&alice, AuthnState::Gone);
    let pending = plane.elect(&doc).unwrap();
    println!("elect started: {} (owner still alice)", pending.as_reason());
    assert_eq!(plane.object(&doc).unwrap().owner, alice);

    match plane.commit_elect(&doc) {
        Err(VerbError::ElectWaitNotElapsed(_)) => println!("commit before wait: refused"),
        other => panic!("expected wait, got {other:?}"),
    }

    plane.set_now(sociacl_core::Timestamp(
        plane.now().0 + plane.elect_wait().0,
    ));
    plane.cancel_will(&doc, &executor).unwrap();
    match plane.commit_elect(&doc) {
        Err(VerbError::WillCanceled(_)) => {
            println!("executor canceled; owner is still alice")
        }
        other => panic!("expected canceled, got {other:?}"),
    }

    plane.elect(&desk).unwrap();
    plane.set_now(sociacl_core::Timestamp(
        plane.now().0 + plane.elect_wait().0,
    ));
    let installed = plane.commit_elect(&desk).unwrap();
    println!("commit after wait: {}", installed.as_reason());
    let check = plane
        .check_named("read", &desk, &bob, PredicateId::owner())
        .unwrap();
    println!("bob owns desk: allowed={}", check.allowed);
}
